semaphore.go

```diff
@@ -0,0 +1,58 @@
+package influxdb
+
+import (
+	"context"
+	"time"
+)
+
+// DefaultLeaseTTL is used when a specific lease TTL is not requested.
+const DefaultLeaseTTL = time.Minute
+
+// A Semaphore provides an API for requesting ownership of an expirable semaphore.
+//
+// Acquired semaphores have an expiration period. If they're not released or extended
+// during this period then they will expire and ownership of the semaphore will
+// be lost.
+type Semaphore interface {
+	// TODO(edd): add Acquire and AcquireTTL when needed. These should block.
+
+	// TryAcquire attempts to acquire ownership of the semaphore. TryAcquire
+	// must not block. Failure to get ownership of the semaphore should be
+	// signalled to the caller via the return of a nil Lease.
+	TryAcquire(context.Context) (Lease, error)
```

Comment:
I would personally use an error to signal that the acquisition failed (e.g. `ErrNoAcquire`). I find it's easy for people to forget to check a `nil` `Lease`.

Reply:
Also, if this is meant to support distributed use cases, it seems like all acquisitions should require a TTL (i.e. only implement `TryAcquireTTL()`).

Reply:
I'm open to doing that. We could probably remove `TryAcquire` since in the end it's unlikely to get many callers (mostly you want to dictate the TTL upfront I guess).